File: app/src/main/java/com/jbrunton/mymovies/app/shared/BaseFragment.java
package com.jbrunton.mymovies.app.shared;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;

import com.jbrunton.mymovies.MyMoviesApplication;
import com.jbrunton.mymovies.app.ApplicationDependencies;

public abstract class BaseFragment<T extends BaseViewModel> extends Fragment {
    private T viewModel;

    @Override public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        viewModel = provideViewModel();
        viewModel.start();
    }

    protected T viewModel() {
        return this.viewModel;
    }

    protected abstract T provideViewModel();

    protected ApplicationDependencies dependencies() {
        return ((MyMoviesApplication) getActivity().getApplication()).dependencies();
    }
}

File: app/src/main/java/com/jbrunton/mymovies/shared/BaseFragment.kt
package com.jbrunton.mymovies.shared

import androidx.lifecycle.ViewModel
import com.jbrunton.mymovies.MyMoviesApplication
import kotlin.reflect.KClass

abstract class BaseFragment<T : BaseViewModel> : androidx.fragment.app.Fragment() {
    fun <T : Any> resolve(klass: KClass<T>): T {
        return (context as BaseActivity<*>).resolve(klass)
    }

    inline fun <reified T: ViewModel> resolve(): T {
        return resolve(T::class)
    }

    inline fun <reified T: ViewModel> resolveViewModel(): T {
        return (context as BaseActivity<*>).resolveViewModel()
    }
}
